### PR TITLE
Add Alibaba Cloud Provider support

### PR DESCRIPTION
**Pull request by ringtail**
Alibaba Cloud Container Service is a high-performance and scalable container application management service that enables you to use Docker and Kubernetes to manage the lifecycle of containerized applications. 

This commit add Alibaba Cloud Provider for cluster autoscaler.

Signed-off-by: ringtail zhongwei.lzw@alibaba-inc.com

### CODE REVIEW

**Comment by warmchang** [comment]
A huge PR. Includes three commits: code, documentation and examples. :+1: 

cc @zhangxiaoyu-zidif

**Comment by aleksandra-malinowska** [comment]
Hi, thanks for making this PR. To make reviewing it easier, can you please split the commit containing cloud provider code into adding dependencies (can be several commits, e.g. separate one for changes in _override and regular Godeps sync) and changes to cloudprovider/?

**Comment by mwielgus** [comment]
Cluster Autoscaler dependency model assumes that all (or 99.99%) of dependencies come directly from kubernetes/kubernetes to avoid problems with library versions, licenses, and security reviews of the dependencies. Please see FAQ for details:

https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#how-can-i-update-ca-dependencies-particularly-k8siokubernetes

Unfortunately, the Alibaba Cloud libraries are not included in kubernetes/kubernetes dependencies. Thus we are unable to accept the addition of your cloud provider to kubernetes/autoscaler. Moreover, we don't have knowledge and capacity required to review the code related to Alibaba Cloud.

What I can recommend is making your own fork of the kubernetes/autoscaler repository and adding your cloud provider there. Assuming that you won't be changing the core code, but instead contributing the improvements to the core repository, the subsequent syncs with the original repository should be relatively painless. You can also import the CA code as a library and call it from your own clone of main.go.

On our side, we can put a link to your repo in the documentation and make sure that the Cluster Autoscaler support for Alibaba Cloud is well advertised.

**Comment by zhangxiaoyu-zidif** [comment]
looks good to me, but  I have no env to validate those.

File: cluster-autoscaler/cloudprovider/alicloud/alicloud_auto_scaling_group.go [hunk 1]
```diff
@@ -0,0 +1,204 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
Comment (zhangxiaoyu-zidif) [inline cluster-autoscaler/cloudprovider/alicloud/alicloud_auto_scaling_group.go]: 2016 ?

**Comment by ringtail** [comment]
@zhangxiaoyu-zidif I can provide you a kubernetes cluster in Alibaba Cloud.Could you get me a more real time communication way so that I can help you to test cluster autoscaler in time.

**Comment by ringtail** [comment]
@mwielgus Sorry,I don't fully understand.Is it due to the dependencies(github.com/aliyun/alibaba-cloud-sdk-go)  updating? If so, I think i can follow the way of GKE Provider extracting the dependencies to the provider instead of the vendor. Fork a custom repository may not be a good choice. kubernetes have accepted cloud-provider-alibaba-cloud already. we are going to cover the other components left. cluster autoscaler is one of the most used components. We hope our customers can use the official releases with latest changes. Thanks!

File: cluster-autoscaler/cloudprovider/alicloud/alicloud_auto_scaling_group.go (hunk 1 above)
Comment (ringtail) [reply cluster-autoscaler/cloudprovider/alicloud/alicloud_auto_scaling_group.go]: I'll change the Copyright to the latest version.

**Comment by zhangxiaoyu-zidif** [comment]
> @zhangxiaoyu-zidif I can provide you a kubernetes cluster in Alibaba Cloud.Could you get me a more real time communication way so that I can help you to test cluster autoscaler in time.

Thanks, I have interest in it. you can mail me though mailbox: zxysjtu@gmail.com

**Comment by warmchang** [comment]
> Thanks, I have interest in it. you can mail me though mailbox: zxysjtu@gmail.com

Hi, You two can communicate directly by intranet or "DingTalk". :smile: 
And aotify the community of the final test steps and conclusions.

**Comment by duffqiu** [comment]
> Cluster Autoscaler dependency model assumes that all (or 99.99%) of dependencies come directly from kubernetes/kubernetes to avoid problems with library versions, licenses, and security reviews of the dependencies. Please see FAQ for details:
> 
> https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#how-can-i-update-ca-dependencies-particularly-k8siokubernetes
> 
> Unfortunately, the Alibaba Cloud libraries are not included in kubernetes/kubernetes dependencies. Thus we are unable to accept the addition of your cloud provider to kubernetes/autoscaler. Moreover, we don't have knowledge and capacity required to review the code related to Alibaba Cloud.
> 
> What I can recommend is making your own fork of the kubernetes/autoscaler repository and adding your cloud provider there. Assuming that you won't be changing the core code, but instead contributing the improvements to the core repository, the subsequent syncs with the original repository should be relatively painless. You can also import the CA code as a library and call it from your own clone of main.go.
> 
> On our side, we can put a link to your repo in the documentation and make sure that the Cluster Autoscaler support for Alibaba Cloud is well advertised.

I seems aws/azure also include their lib to support autoscaler. why Alibaba cloud can't do the same way?
image

**Comment by mwielgus** [comment]
@duffqiu
The reason AWS, GCE and Azure are there is that the corresponding cloud provider libraries are a part of the default Kubernetes distribution, thus reusing the dependencies from the main Kubernetes repository is possible and there are (usually) no sync conflicts. Alibaba cloud is not included in the default Kubernetes repository.

See: https://github.com/kubernetes/kubernetes/tree/master/pkg/cloudprovider/providers

I guess there is above 20 cloud providers who offer some Kubernetes integration. We cannot have drivers for all of them (and the following) in the main repo.  Azure, AWS, GCE were lucky to be added early in the process:

* https://github.com/kubernetes/contrib/pull/1377
* https://github.com/kubernetes/contrib/pull/2146

For cloud providers who joined later we have the following options:

* Wait for/contribute to ClusterAPI which will abstract cloud providers.
* Fork
* Include CA as a library in your repo.
* Get your cloud provider approved to core Kubernetes.

**Comment by mwielgus** [comment]
Closing in favour of  #1309.

**Pull request closed without merging**